stats/stats.go

```diff
@@ -55,52 +55,72 @@ func ConditionallyStartStatsServer(addhandlers ...func(mux *http.ServeMux, profi
 }
 
 func StartStatsServerWithPort(startupOpts StartupOptions, addhandlers ...func(mux *http.ServeMux, profiles map[string]string)) {
+	StartCancellableStatsServerWithPort(context.Background(), startupOpts, addhandlers...)
+}
+
+func StartCancellableStatsServerWithPort(ctx context.Context, startupOpts StartupOptions, customAddHandlers ...func(mux *http.ServeMux, profiles map[string]string)) {
 	// if the env var was provided (i.e., startup is conditional) and the value of that env var is not the expected value, then return and do nothing
 	if startupOpts.EnvVar != "" && os.Getenv(startupOpts.EnvVar) != startupOpts.EnabledValue {
 		return
 	}
 
-	var logLevel zap.AtomicLevel
-	if envLogLevel := os.Getenv("LOG_LEVEL"); envLogLevel != "" {
-		var setLevel zapcore.Level
-		switch envLogLevel {
-		case "debug":
-			setLevel = zapcore.DebugLevel
-		case "warn":
-			setLevel = zapcore.WarnLevel
-		case "error":
-			setLevel = zapcore.ErrorLevel
-		case "panic":
-			setLevel = zapcore.PanicLevel
-		case "fatal":
-			setLevel = zapcore.FatalLevel
-		default:
-			setLevel = zapcore.InfoLevel
-		}
+	if envLogLevel := os.Getenv(contextutils.LogLevelEnvName); envLogLevel != "" {
+		contextutils.SetLogLevelFromString(envLogLevel)
+	}
 
-		contextutils.SetLogLevel(setLevel)
+	go RunCancellableGoroutineStat(ctx)
 
-	} else if startupOpts.LogLevel != nil {
-		logLevel = *startupOpts.LogLevel
-	}
+	// The running instance of the Stats server
+	var server *http.Server
 
-	go RunGoroutineStat()
+	addHandlers := append(customAddHandlers, addPprof, addStats)
 
+	// Run the server in a goroutine
 	go func() {
 		mux := new(http.ServeMux)
 
-		mux.Handle("/logging", logLevel)
+		mux.Handle("/logging", getLoggingHandler(startupOpts))
 
-		addhandlers = append(addhandlers, addPprof, addStats)
-
-		for _, addhandler := range addhandlers {
-			addhandler(mux, profileDescriptions)
+		for _, addHandler := range addHandlers {
+			addHandler(mux, profileDescriptions)
 		}
 
 		// add the index
 		mux.HandleFunc("/", Index)
-		http.ListenAndServe(fmt.Sprintf(":%d", startupOpts.Port), mux)
+
+		server = &http.Server{
+			Addr:    fmt.Sprintf(":%d", startupOpts.Port),
+			Handler: mux,
+		}
+		err := server.ListenAndServe()
+		if err == http.ErrServerClosed {
+			contextutils.LoggerFrom(ctx).Infof("Stats server closed")
+		} else {
+			contextutils.LoggerFrom(ctx).Warnf("Stats server closed with unexpected error: %v", err)
+		}
 	}()
+
+	// Run a separate goroutine to handle the server shutdown when the context is cancelled
+	go func() {
+		select {
+		case <-ctx.Done():
```

Comment:
nit, no need to select with a single case. we can just do `<-ctx.Done()`
```suggestion
		<-ctx.Done()
````